tokio/src/sync/watch.rs

```diff
@@ -610,6 +610,53 @@ impl<T> Receiver<T> {
         }
     }
 
+    /// Takes a closure and waits until the provided closure returns true.
+    ///
+    /// This function is similar to [`Receiver::changed()`], but it takes a closure that is
+    /// called with a reference to the new value. If the closure returns `true`,
+    /// then the function returns immediately. Otherwise, it waits for a new
+    /// value and calls the closure again.
+    ///
+    /// [`Receiver::changed()`]: crate::sync::watch::Receiver::changed
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::watch;
+    ///
+    /// #[tokio::main]
+    ///
+    /// async fn main() {
+    ///     let (tx, _rx) = watch::channel("hello");
+    ///
+    ///     tx.send("goodbye").unwrap();
+    ///     
+    ///     // here we subscribe to a second receiver
+    ///     // now in case of using `changed` we would have
+    ///     // to first check the current value and then wait
+    ///     // for changes or else `changed` would hang.
+    ///     let mut rx2 = tx.subscribe();
+    ///     
+    ///     // in place of changed we have use `wait_for`
+    ///     // which would automatically check the current value
+    ///     // and wait for changes until the closure returns true.
+    ///     assert!(rx2.wait_for(|val| *val == "goodbye").await.is_ok());
+    ///     assert_eq!(*rx2.borrow(), "goodbye");
+    /// }
+    /// ```
+    pub async fn wait_for(
+        &mut self,
+        mut f: impl FnMut(&T) -> bool,
+    ) -> Result<(), error::RecvError> {
```

Comment:
I think the main remaining question is whether this should return `()` or a reference to the item on success.
```suggestion
    pub async fn wait_for<'a>(
        &'a mut self,
        mut f: impl FnMut(&T) -> bool,
    ) -> Result<Ref<'a, T>, error::RecvError> {
```
@carllerche What do you think?